python/revng/scripting/projects/project.py

```diff
@@ -0,0 +1,151 @@
+        pass
+
+    @abstractmethod
+    def analyses_list(self, analysis_name: str) -> None:
+        pass
+
+    @abstractmethod
+    def commit(self) -> None:
+        pass
+
+    def revert(self) -> None:
+        diff = make_diff(self.model, self.last_saved_model)
+        if not diff.changes:
+            return
+        self.set_model(deepcopy(self.last_saved_model))
+
+    @abstractmethod
+    def _get_pipeline_description(self) -> None:
+        pass
+
+    def import_and_analyze(self, input_binary_path: str):
+        self.import_binary(input_binary_path)
+        self.analyses_list("revng-initial-auto-analysis")
+        self.analyses_list("revng-c-initial-auto-analysis")
+        self._get_pipeline_description()
+        self.set_model_mixins()
+
+    def set_model(self, model: Binary) -> None:
+        self.model = model
+        self.last_saved_model = deepcopy(self.model)
+        # We call this method in multiple places, during the initial auto
+        # analysis for example we don't have the pipeline description yet
+        # so we need to take that into account.
+        if self.pipeline_description:
+            self.set_model_mixins()
+
+    def set_model_mixins(self) -> None:
+        for ct in self.pipeline_description["Ranks"]:
+            model_path = ct.get("ModelPath")
+            if not model_path:
+                continue
+
+            obj = get_element_by_path(model_path, self.model)
+            if isinstance(obj, TypedList):
+                for elem in obj:
+                    elem._project = self
+            else:
+                obj._project = self
+
+    def _get_analysis_inputs(self, analysis_name: str) -> dict:
+        inputs: dict[str, List[str]] = {}
+        for step in self.pipeline_description["Steps"]:
+            for analysis in step["Analyses"]:
+                if analysis["Name"] == analysis_name:
+                    for i in analysis["ContainerInputs"]:
+                        inputs[i["Name"]] = i["AcceptableKinds"]
+        return inputs
+
+    def _get_artifact_kind(self, step_name: str) -> str:
+        for step in self.pipeline_description["Steps"]:
+            if step["Name"] == step_name:
+                return step["Artifacts"]["Kind"]
+        raise RuntimeError(f"Couldn't find step: {step_name}")
+
+    def _get_artifact_container(self, step_name) -> str:
+        for step in self.pipeline_description["Steps"]:
+            if step["Name"] == step_name:
+                return step["Artifacts"]["Container"]
+        raise RuntimeError(f"Couldn't find step: {step_name}")
+
+    def _get_step_name(self, analysis_name: str) -> str:
+        for step in self.pipeline_description["Steps"]:
+            for analysis in step["Analyses"]:
+                if analysis["Name"] == analysis_name:
+                    return step["Name"]
+        raise RuntimeError(f"Couldn't find step for analysis: {analysis_name}")
+
+    def _get_analyses_list_names(self) -> list[str]:
```

Comment:
`list` should not be used anywhere, double check that you're not using feature that are not available in python 3.8